File: arrays/cpp/001_twosum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int>& nums, int target) {  
    unordered_map<int, int> ump;
    int n = nums.size();

    for(int i = 0; i < n; i++) {
      if (ump.find(target-nums[i]) != ump.end()) {
        return vector<int> ({ump[target-nums[i]], i});
      }
      ump[nums[i]] = i;
    }

    return vector<int> ({-1, -1});
  }
};

File: arrays/python/001_two_sum.py
# O(n) solution
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_dict = {}

        for i in range(len(nums)):
            nums_dict[nums[i]] = i

        for i in range(len(nums)):
            num_diff = target - nums[i]
            if (num_diff in nums_dict): 
                if (nums_dict[num_diff] != i):
                    return [i, nums_dict[num_diff]]


# O(n^2) solution
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if (nums[i]+nums[j]) == target:
                    return [i, j]
